src/functional/ffront/type_info.py

```diff
@@ -1,48 +1,235 @@
-        case DeferredSymbolType():
+    match symbol_type:
+        case ct.DeferredSymbolType():
             return False
-        case SymbolType():
+        case ct.SymbolType():
             return True
     return False
 
 
-def is_complete_scalar_type(sym_type: Optional[SymbolType]) -> TypeGuard[ScalarType]:
-    if isinstance(sym_type, ScalarType):
+class TypeKind(enum.Enum):
+    FIELD = 0
+    SCALAR = 1
+    UNKNOWN = 2
+
+
+def type_kind(symbol_type: ct.SymbolType) -> TypeKind:
+    """
+    Determine whether ``symbol_type`` is scalar, field or unknown.
+
+    Examples:
+    ---------
+    >>> type_kind(ct.DeferredSymbolType(constraint=None)).name
+    'UNKNOWN'
+
+    >>> type_kind(ct.ScalarType(kind=ct.ScalarKind.BOOL)).name
+    'SCALAR'
+
+    >>> type_kind(ct.FunctionType(args=[], kwargs={}, returns=ct.FieldType(dims=[], dtype=ct.ScalarType(kind=ct.ScalarKind.INT32)))).name
+    'FIELD'
+
+    >>> type_kind(ct.TupleType(types=[ct.ScalarType(kind=ct.ScalarKind.FLOAT64)])).name
+    'SCALAR'
+    """
+    match symbol_type:
+        case ct.DeferredSymbolType(constraint=None):
+            return TypeKind.UNKNOWN
+        case ct.TupleType(types=subtypes):
+            if subtypes:
+                return type_kind(subtypes[0])
+        case ct.FunctionType(returns=returntype):
+            return type_kind(returntype)
+
+    match type_class(symbol_type):
+        case ct.FieldType:
+            return TypeKind.FIELD
+        case ct.ScalarType:
+            return TypeKind.SCALAR
+
+    return TypeKind.UNKNOWN
+
+
+def type_class(symbol_type: ct.SymbolType) -> Type[ct.SymbolType]:
+    """
+    Determine which class should be used to create a compatible concrete type.
+
+    Examples:
+    ---------
+    >>> type_class(ct.DeferredSymbolType(constraint=ct.ScalarType)).__name__
+    'ScalarType'
+
+    >>> type_class(ct.FieldType(dims=[], dtype=ct.ScalarType(kind=ct.ScalarKind.BOOL))).__name__
+    'FieldType'
+
+    >>> type_class(ct.TupleType(types=[])).__name__
+    'TupleType'
+    """
+    match symbol_type:
+        case ct.DeferredSymbolType(constraint):
+            if constraint is None:
+                raise GTTypeError(f"No type information available for {symbol_type}!")
+            return constraint
+        case ct.SymbolType() as concrete_type:
+            return concrete_type.__class__
+    raise GTTypeError(
+        f"Invalid type for TypeInfo: requires {ct.SymbolType}, got {type(symbol_type)}!"
+    )
+
+
+def extract_dtype(symbol_type: ct.SymbolType) -> ct.ScalarType:
+    """
+    Extract the data type from ``symbol_type`` if possible.
+
+    Examples:
+    ---------
+    >>> print(extract_dtype(ct.ScalarType(kind=ct.ScalarKind.FLOAT64)))
+    float64
+
+    >>> print(extract_dtype(ct.FieldType(dims=[], dtype=ct.ScalarType(kind=ct.ScalarKind.BOOL))))
+    bool
+
+    >>> print(extract_dtype(ct.FunctionType(args=[], kwargs={}, returns=ct.ScalarType(kind=ct.ScalarKind.INT32, shape=[2, 2]))))
+    int32[2, 2]
```

Comment:
I do not agree with the dtype of a function being the dtype of its return value, I believe that is a different question to ask.

Reply:
What other dtype would you extract from a function type? The dtype of a random argument?

Reply:
It comes down to "Best effort" vs "Defensive". If the function should provide a best effort result, it is unambiguously expected to return the return value dtype. If the function should be defensive it should fail in as many cases as can reasonably argued for. I prefer best effort and will document the behaviour in the docstring.

Reply:
> I do not agree with the dtype of a function being the dtype of its return value, I believe that is a different question to ask.

Nothing here says that the dtype of a function is the dtype of it's return value. It sais if you *extract* a dtype from a function, what you are really asking must be to *extract* the dtype from it's return type. Anything else would not make sense.